pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,1094 @@
+	}
+	return mutated, nil
+}
+
+func deleteFromMap(original *libovsdb.OvsMap, toDelete interface{}) (*libovsdb.OvsMap, error) {
+	mutated := new(libovsdb.OvsMap)
+	copier.Copy(&mutated, &original)
+	switch toDelete := toDelete.(type) {
+	case libovsdb.OvsMap:
+		for k, v := range toDelete.GoMap {
+			if mutated.GoMap[k] == v {
+				delete(mutated.GoMap, k)
+			}
+		}
+	case libovsdb.OvsSet:
+		for _, k := range toDelete.GoSet {
+			delete(mutated.GoMap, k)
+		}
+	}
+	return mutated, nil
+}
+
+func (m *Mutation) MutateMap(row *map[string]interface{}) error {
+	original := (*row)[m.Column].(libovsdb.OvsMap)
+	mutated := new(libovsdb.OvsMap)
+	var err error
+	switch m.Mutator {
+	case MT_INSERT:
+		mutated, err = insertToMap(&original, m.Value)
+	case MT_DELETE:
+		mutated, err = deleteFromMap(&original, m.Value)
+	default:
+		err = errors.New(E_CONSTRAINT_VIOLATION)
+	}
+	if err != nil {
+		return err
+	}
+	(*row)[m.Column] = mutated
+	return nil
+}
+
+func (m *Mutation) Mutate(row *map[string]interface{}) error {
+	switch m.Column {
+	case COL_UUID, COL_VERSION:
+		return errors.New(E_CONSTRAINT_VIOLATION)
+	}
+	switch (*row)[m.Column].(type) {
+	case int:
+		return m.MutateInteger(row)
+	case float64:
+		return m.MutateReal(row)
+	case libovsdb.OvsSet:
+		return m.MutateSet(row)
+	case libovsdb.OvsMap:
+		return m.MutateMap(row)
+	default:
+		return errors.New(E_CONSTRAINT_VIOLATION)
+	}
+}
+
+func RowMutate(original *map[string]interface{}, mutations []interface{}) error {
+	// XXX(alexey): can we just run on original ?
+	mutated := &map[string]interface{}{}
+	copier.Copy(mutated, original)
+	for _, mt := range mutations {
+		mutation, err := NewMutation(mt.([]interface{}))
+		if err != nil {
+			return err
+		}
+		err = mutation.Mutate(mutated)
+		if err != nil {
+			return err
+		}
+	}
+	copier.Copy(original, mutated)
+	return nil
+}
+
+func RowUpdate(original *map[string]interface{}, update map[string]interface{}) error {
+	for k, v := range update {
+		switch k {
+		case COL_UUID, COL_VERSION:
+			return errors.New(E_CONSTRAINT_VIOLATION)
+		}
+		(*original)[k] = v
+	}
+	return nil
+}
+
+func etcdGetTable(txn *Transaction, table string) {
+	prefix := makePrefix(txn.request.DBName, table)
+	etcdOp := clientv3.OpGet(prefix, clientv3.WithPrefix())
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+}
+
+func etcdGetRow(txn *Transaction, table, uuid string) {
+	key := makeKey(txn.request.DBName, table, uuid)
+	etcdOp := clientv3.OpGet(key)
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+}
+
+func etcdGetTableOrRow(txn *Transaction, table, uuid string) {
+	if uuid != "" {
+		etcdGetRow(txn, table, uuid)
+	} else {
+		etcdGetTable(txn, table)
+	}
+}
+
+func etcdGetByWhere(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	uuid, err := doesWhereContainCondTypeUUID(ovsOp.Where)
+	if err != nil {
+		ovsResult.Error = err.Error()
+		return err
+	}
+	etcdGetTableOrRow(txn, ovsOp.Table, uuid)
```

Comment:
if you have 2 operations on the same table, will you bring it twice ?

Reply:
currently yes, should be merged into one call